### PR TITLE
remove Env type and Data trait

### LINKED ISSUE #8: Remove Env and Data
The Env class and the Data trait are vestigial by this point in the Masonry codebase.

There is currently no support for updating widgets when the Env is changed, and I have no intention of adding that support (Xilem [seems to be moving in the same direction](https://github.com/linebender/xilem/issues/7)). This means Env is basically just a way to pass a singleton around in Widget code.

We should remove Env and data, and just hardcode the values that we currently get from Env, with future plans to pass them as arguments or something. Frontend frameworks might still use something Env-like to get style data, but that'll be their responsibility.

### LINKED ISSUE #373: Remove Env and Data
The Env class and the Data trait are vestigial by this point in the Masonry codebase.

There is currently no support for updating widgets when the Env is changed, and I have no intention of adding that support (Xilem [seems to be moving in the same direction](https://github.com/linebender/xilem/issues/7)). This means Env is basically just a way to pass a singleton around in Widget code.

We should remove Env and data, and just hardcode the values that we currently get from Env, with future plans to pass them as arguments or something. Frontend frameworks might still use something Env-like to get style data, but that'll be their responsibility.

### PR DESCRIPTION
**Pull request by herkhinah**
I've refactored out the Env type and Data trait. The crate compiles and the examples run. Fixes linebender/xilem#373

### CODE REVIEW

**Comment by PoignardAzur** [comment]
Sorry for the late reply, I've been on vacation. I've only had time to skim this, but it looks like a clean refactor. Pretty high-quality contribution!

I've noticed a `// TODO remove env` in `widget/flex.rs` that you probably didn't mean to leave.

I'll take another look at this soon, and after that, we'll try to get this PR merged ASAP.

File: src/widget/flex.rs [hunk 1]
```diff
@@ -693,11 +685,13 @@ impl Widget for Flex {
...
+    fn paint(&mut self, ctx: &mut PaintCtx) {
         for child in self.children.iter_mut().filter_map(|x| x.widget_mut()) {
-            child.paint(ctx, env);
+            child.paint(ctx);
         }
 
+        // TODO remove env
+        /*
         // paint the baseline if we're debugging layout
         if env.get(Env::DEBUG_PAINT) && ctx.widget_state.baseline_offset != 0.0 {
```
Comment (PoignardAzur) [inline src/widget/flex.rs:696]: The refactor needs to address these kinds of cases. I'd suggest adding `debug_paint`, `debug_widget` and `debug_widget_id` bool attributes to GlobalPassCtx and going from there.

File: src/widget/widget_pod.rs [hunk 2]
```diff
@@ -258,13 +254,16 @@ impl<W: Widget> WidgetPod<W> {
                 widget_state: inner_state,
             };
 
+            // TODO Env removal
+            /*
```
Comment (PoignardAzur) [inline src/widget/widget_pod.rs:258]: See above.

File: src/widget/widget_pod.rs [hunk 3]
```diff
@@ -1038,13 +1036,15 @@ impl<W: Widget> WidgetPod<W> {
     /// [`layout`]: trait.Widget.html#tymethod.layout
     /// [`Widget::paint`]: trait.Widget.html#tymethod.paint
     /// [`paint`]: #method.paint
-    pub fn paint_raw(&mut self, ctx: &mut PaintCtx, env: &Env) {
+    pub fn paint_raw(&mut self, ctx: &mut PaintCtx) {
         self.mark_as_visited();
 
+        // TODO env removal
+        /*
```
Comment (PoignardAzur) [inline src/widget/widget_pod.rs:1043]: See above.

File: src/widget/widget_pod.rs [hunk 4]
```diff
@@ -1057,17 +1057,20 @@ impl<W: Widget> WidgetPod<W> {
                 region: ctx.region.clone(),
                 depth: ctx.depth,
             };
-            widget_pod.inner.paint(&mut inner_ctx, env);
+            widget_pod.inner.paint(&mut inner_ctx);
 
+            // TODO: Env removal
+            /*
```
Comment (PoignardAzur) [inline src/widget/widget_pod.rs:1063]: See above.

File: src/text/rich_text.rs [hunk 5]
```diff
@@ -23,6 +25,14 @@ pub struct RichText {
     links: Arc<[Link]>,
 }
 
+impl PartialEq for RichText {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.buffer, &other.buffer) && Arc::ptr_eq(&self.attrs, &other.attrs) && Arc::ptr_eq(&self.links, &other.links)
+    }
+}
+
+impl Eq for RichText {}
```
Comment (PoignardAzur) [inline src/text/rich_text.rs:34]: This is an invalid implementation of `Eq` and a dubious implementation of `PartialEq`. Two buffers with different pointer values but identical content shouldn't return false in PartialEq.

As a placeholder solution, I'd suggest adding a function called eg `maybe_eq` to the TextStorage trait with the same semantic as the now-removed `Data::same()`. We can think of a better design once this is merged.

**Review by PoignardAzur — CHANGES_REQUESTED** [review]
Overall this seems like a clean set of changes. Once the items I mentioned are addressed, and the rustfmt and clippy errors are fixed, we can merge this ASAP.

By the way, please make sure to do any followup fixes in a separate commit (eg don't use rebase or `git commit --amend`) so the fixes don't get lost in this PR's huge changeset.

**Comment by PoignardAzur** [comment]
FYI, I might rebase this PR on my side (or straight-up redo it from scratch) next week, since I'm actively working on Masonry now.

**Comment by PoignardAzur** [comment]
I've done a bunch of changes on top of this PR and merged them. Thanks for the contribution!

**Comment by herkhinah** [comment]
Sry for not responding. I’m currently homeless for an unknown amount of time.Sent from my iPhoneOn 26. Mar 2024, at 15:25, Olivier FAURE ***@***.***> wrote:﻿
I've done a bunch of changes on top of this PR and merged them. Thanks for the contribution!

**Comment by PoignardAzur** [comment]
Shit. Good luck then.

**Pull request closed without merging**